pkg/blockbuilder/builder/slimgester.go

```diff
@@ -95,94 +111,170 @@ type BlockBuilder struct {
 }
 
 func NewBlockBuilder(
 	id string,
 	cfg Config,
 	periodicConfigs []config.PeriodConfig,
+	readerFactory func(partition int32) (partition.Reader, error),
 	store stores.ChunkWriter,
 	objStore *MultiStore,
 	logger log.Logger,
 	reg prometheus.Registerer,
-	jobController *PartitionJobController,
 ) (*BlockBuilder,
 	error) {
+	decoder, err := kafka.NewDecoder()
+	if err != nil {
+		return nil, err
+	}
+
 	i := &BlockBuilder{
 		id:              id,
 		cfg:             cfg,
 		periodicConfigs: periodicConfigs,
 		metrics:         NewSlimgesterMetrics(reg),
 		logger:          logger,
+		decoder:         decoder,
+		readerFactory:   readerFactory,
 		store:           store,
 		objStore:        objStore,
-		jobController:   jobController,
+		// TODO: wire transport
 	}
 
 	i.Service = services.NewBasicService(nil, i.running, nil)
 	return i, nil
 }
 
 func (i *BlockBuilder) running(ctx context.Context) error {
-	ticker := time.NewTicker(i.cfg.Interval)
-	defer ticker.Stop()
+	wg := sync.WaitGroup{}
 
-	// run once in beginning
-	select {
-	case <-ctx.Done():
-		return nil
-	default:
-		_, err := i.runOne(ctx)
-		if err != nil {
-			level.Error(i.logger).Log("msg", "block builder run failed", "err", err)
-		}
+	for j := 0; j < i.cfg.WorkerParallelism; j++ {
+		wg.Add(1)
+		go func(id string) {
+			defer wg.Done()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					err := i.runOne(ctx, id)
+					if err != nil {
+						level.Error(i.logger).Log("msg", "block builder run failed", "err", err)
+					}
+				}
+			}
+		}(fmt.Sprintf("worker-%d", j))
 	}
 
-	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		case <-ticker.C:
-			skipped, err := i.runOne(ctx)
-			level.Info(i.logger).Log(
-				"msg", "completed block builder run", "skipped",
-				"skipped", skipped,
-				"err", err,
-			)
-			if err != nil {
-				level.Error(i.logger).Log("msg", "block builder run failed", "err", err)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		ticker := time.NewTicker(i.cfg.SyncInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				if err := i.syncJobs(ctx); err != nil {
+					level.Error(i.logger).Log("msg", "failed to sync jobs", "err", err)
+				}
 			}
 		}
-	}
+	}()
+
+	wg.Wait()
+	return nil
 }
 
-// runOne performs a single
-func (i *BlockBuilder) runOne(ctx context.Context) (skipped bool, err error) {
+func (i *BlockBuilder) syncJobs(ctx context.Context) error {
+	i.jobsMtx.RLock()
+	defer i.jobsMtx.RUnlock()
+
+	for _, job := range i.inflightJobs {
+		if err := i.transport.SendSyncJob(ctx, &types.SyncJobRequest{
+			BuilderID: i.id,
+			Job:       job,
+		}); err != nil {
+			level.Error(i.logger).Log("msg", "failed to sync job", "err", err)
+		}
+	}
+
+	return nil
+}
 
-	exists, job, err := i.jobController.LoadJob(ctx)
+func (i *BlockBuilder) runOne(ctx context.Context, workerID string) error {
+	// assuming GetJob blocks/polls until a job is available
```

Comment:
I suspect we'll need to retry when there are no jobs here, but as you said it's also possible the transport handles this